cryptol-remote-api/python/cryptol/synchronous.py

```diff
@@ -0,0 +1,283 @@
+
+    def check(self, expr : Any, *, num_tests : Union[Literal['all'], int, None] = None) -> CheckReport:
+        """Tests the validity of a Cryptol expression with random inputs. The expression must be a function with
+        return type ``Bit``.
+        If ``num_tests`` is ``"all"`` then the expression is tested exhaustively (i.e., against all possible inputs).
+        If ``num_tests`` is omitted, Cryptol defaults to running 100 tests.
+        """
+        return to_check_report(self.connection.check_raw(expr, num_tests=num_tests).result())
+
+    def check_type(self, code : Any) -> cryptoltypes.CryptolType:
+        """Check the type of a Cryptol expression, represented according to
+        :ref:`cryptol-json-expression`, with Python datatypes standing for
+        their JSON equivalents.
+        """
+        return cryptoltypes.to_type(self.connection.check_type(code).result()['type'])
+
+    @overload
+    def sat(self, expr : Any, solver : OnlineSolver = Z3, count : int = 1) -> Union[Satisfiable, Unsatisfiable]: ...
+    @overload
+    def sat(self, expr : Any, solver : OfflineSolver, count : int = 1) -> OfflineSmtQuery: ...
+
+    def sat(self, expr : Any, solver : Solver = Z3, count : int = 1) -> Union[Satisfiable, Unsatisfiable, OfflineSmtQuery]:
+        """Check the satisfiability of a Cryptol expression, represented according to
+        :ref:`cryptol-json-expression`, with Python datatypes standing for
+        their JSON equivalents. Use the solver named `solver`, and return up to
+        `count` solutions.
+        """
+        if isinstance(solver, OfflineSolver):
+            res = self.connection.sat_raw(expr, solver, count).result()
+            if res['result'] == 'offline':
+                return OfflineSmtQuery(res['query'])
+            else:
+                raise ValueError("Expected an offline SMT result, got: " + str(res))
+        elif isinstance(solver, OnlineSolver):
+            res = self.connection.sat_raw(expr, solver, count).result()
+            if res['result'] == 'unsatisfiable':
+                return Unsatisfiable()
+            elif res['result'] == 'satisfied':
+                return Satisfiable([[from_cryptol_arg(arg['expr']) for arg in m] for m in res['models']])
+            else:
+                raise ValueError("Unexpected 'sat' SMT result: " + str(res))
+        else:
+            raise ValueError("Unknown solver type: " + str(solver))
+
+    @overload
+    def prove(self, expr : Any, solver : OnlineSolver = Z3) -> Union[Qed, Counterexample]: ...
+    @overload
+    def prove(self, expr : Any, solver : OfflineSolver) -> OfflineSmtQuery: ...
+
+    def prove(self, expr : Any, solver : Solver = Z3) -> Union[Qed, Counterexample, OfflineSmtQuery]:
+        """Check the validity of a Cryptol expression, represented according to
+        :ref:`cryptol-json-expression`, with Python datatypes standing for
+        their JSON equivalents. Use the solver named `solver`.
+        """
```

Comment:
We should add a note in this docstring regarding being able to boolean test `Qed ` and `Counterexample ` (i.e., that the latter is falsy).